File: app/javascript/channels/server_channel.js
// app/javascript/channels/server_channel.js
import { createConsumer } from "@rails/actioncable"

export function initializeServerChat(serverId, currentUserId) {
    const consumer = createConsumer()

    const subscription = consumer.subscriptions.create(
        {
            channel: "ServerChannel",
            server_id: serverId
        },
        {
            connected() {
                console.log(`Connected to server ${serverId} chat`)
            },

            disconnected() {
                console.log(`Disconnected from server ${serverId} chat`)
            },

            received(data) {
                const chatMessages = document.getElementById('server-chat-messages')
                if (chatMessages) {
                    const messageElement = document.createElement('div')
                    messageElement.classList.add('chat-message')
                    messageElement.innerHTML = `
            <span class="message-user">${data.user_id}</span>
            <span class="message-text">${data.message}</span>
            <span class="message-time">${new Date(data.timestamp).toLocaleTimeString()}</span>
          `
                    chatMessages.appendChild(messageElement)
                    chatMessages.scrollTop = chatMessages.scrollHeight
                }
            }
        }
    )

    return {
        sendMessage(message) {
            subscription.send({
                server_id: serverId,
                user_id: currentUserId,
                message: message
            })
        }
    }
}

File: app/channels/server_channel.rb
# app/channels/chat_channel.rb
class ServerChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_#{params[:server_id]}"
  end

  def receive(data)
    ActionCable.server.broadcast(
      "chat_#{params[:server_id]}",
      message: render_message(data['message'], current_user)
    )
  end

  private

  def render_message(message, user)
    ApplicationController.renderer.render(
      partial: 'messages/message',
      locals: { message:, user: }
    )
  end
end